Get track number/length from MusicBrainz where null

scottpk opened the pull request:
Add ability to get track number and length from MusicBrainz for tracks where this is null

**Changes**
<!-- Describe your changes here in 1-5 sentences. -->
Adds the ability to fill track number and length from MusicBrainz for tracks which have this data as null. This looks for null, so that existing metadata will not be overwritten.

![image](https://user-images.githubusercontent.com/1653992/219503483-6b3ef023-b54d-4258-a682-0be385465744.png)

Note that since this is a numeric field, it will be the numeric ordering (as in, for records B-side tracks will be after A-side tracks).

**Issues**
Relates to #4991 Jellyfin doesn't read track numbers from Vinyl records correctly e.g. A1, A2 & B1 etc (user's vinyl software did not fill in the track number)

scottpk committed:
Add ability to get track number and length from MusicBrainz for tracks where this is null

scottpk committed:
Null check to pass the test... previous code worked fine on my computer *shrug*

CluelessTechnologist commented:
Please see discussion in previous PR: https://github.com/jellyfin/jellyfin/pull/7672

So since https://github.com/jellyfin/jellyfin/pull/7603 has been merged I see no blockers to have someone look over this PR? Or are there more rewrites planned to MusicBrainz plugin? @Shadowghost

Shadowghost commented:
Not from my side. This should be good to go if it works as intended

scottpk commented:
Apologies, I am less familiar with git still. How do I request a review?

CluelessTechnologist commented:
Any update on the review? @dmitrylyzo @thornbill 

I understand there might be other more pressing PRs to review. But maybe it's possible to take a quick look and let us know if there's any major issues we need to address? Thanks!

CluelessTechnologist commented:
I see. We need a backend maintainer to review this. Maybe @cvium or @crobibero since they reviewed https://github.com/jellyfin/jellyfin/pull/7603?

cvium commented on MediaBrowser.Providers/Plugins/MusicBrainz/MusicBrainzAlbumProvider.cs:
```diff
@@ -256,6 +256,71 @@ public async Task<MetadataResult<MusicAlbum>> GetMetadata(AlbumInfo info, Cancel
+        {
+            if (result.Item.Children.Any(s => s is Audio && s.IndexNumber == null))
+            {
+                var rspobj = _musicBrainzQuery.LookupRelease(Guid.Parse(releaseId ?? string.Empty), Include.Recordings);
```
String.empty doesn't make much sense here.

scottpk replied:
`String.Empty` is there because the project's StyleCop settings are causing this:

![image](https://github.com/jellyfin/jellyfin/assets/1653992/e16ff020-88a5-42c4-897c-30532c269e4e)

That's the only reason I've done this, it prevents the build otherwise

cvium replied:
It doesn't make sense to look up an empty id. Check that it isn't null before you try to parse it

scottpk committed:
#r1208107071 move null check into if-condition

cvium commented on MediaBrowser.Providers/Plugins/MusicBrainz/MusicBrainzAlbumProvider.cs:
```diff
@@ -256,6 +256,71 @@ public async Task<MetadataResult<MusicAlbum>> GetMetadata(AlbumInfo info, Cancel
             }
         }
 
+        if (Plugin.Instance != null && Plugin.Instance.Configuration.GetMissingTrackInfo)
```
This entire section needs some refactoring. The nested indentation levels are too deep and it performs N + N + 3 database lookups. `result.Item.Children` always hits the db.

cvium replied:
This entire section needs some refactoring. The nested indentation levels are too deep and it performs N + N + 3 database lookups. `result.Item.Children` always hits the db. The last two foreach could also be combined into one.

scottpk committed:
evaluate result.item.Children to an array to avoid extra db calls

collapse nested foreach's and null check's

put runtime update and track index update in same loop

scottpk committed:
compiler does not like having ReadOnlySpan<char> in this foreach declaration, use var and AsSpan() instead

scottpk committed:
use the existing fix

cvium commented on MediaBrowser.Providers/Plugins/MusicBrainz/MusicBrainzAlbumProvider.cs:
```diff
@@ -256,6 +256,54 @@ public async Task<MetadataResult<MusicAlbum>> GetMetadata(AlbumInfo info, Cancel
+                        if (children.Where(p => Guid.Equals(p.Id, c.ParentId)).FirstOrDefault() is MusicAlbum parent)
+                        {
+                            parent.AddChild(c);
+                            result.Item.AddChild(parent);
```
Why is a musicalbum being set as the parent of another musicalbum?

cvium commented on MediaBrowser.Providers/Plugins/MusicBrainz/MusicBrainzAlbumProvider.cs:
```diff
@@ -256,6 +256,54 @@ public async Task<MetadataResult<MusicAlbum>> GetMetadata(AlbumInfo info, Cancel
+
+                    if ((ticks != c.RunTimeTicks) || (index != c.IndexNumber))
+                    {
+                        if (children.Where(p => Guid.Equals(p.Id, c.ParentId)).FirstOrDefault() is MusicAlbum parent)
```
Can this ever happen. A MusicAlbum is not Audio, so there should be no musicalbums in children

scottpk replied:
Yes, because this loop is filtered by instances where `RunTimeTicks` or `IndexNumber` is `null` meaning that this metadata is missing.

cvium commented on MediaBrowser.Providers/Plugins/MusicBrainz/MusicBrainzAlbumProvider.cs:
```diff
@@ -256,6 +256,54 @@ public async Task<MetadataResult<MusicAlbum>> GetMetadata(AlbumInfo info, Cancel
+
+                IList<ITrack> tracks = new List<ITrack>();
+
+                if (rspobj.Media != null)
```
If it's null, is there any reason to continue?

scottpk committed:
Merge branch 'jellyfin:master' into master

scottpk committed:
style: use LookupReleaseAsync
style: do not force IList
style: compare MediaType using == rather than string.Equals

scottpk committed:
style: use compound assignment

style: dispose before new Query

scottpk committed:
Merge branch 'jellyfin:master' into master
